File: backend/src/modules/commerce/cart.service.ts
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import {
  inMemoryStore,
  generateId,
  timestamp,
  CartRecord,
  CartItemRecord,
  ProductRecord,
  ProductVariantRecord,
  CheckoutIntentRecord
} from '../../lib/persistence/in-memory-store';
import { OrdersService } from './orders.service';
import { hashPassword } from '../../lib/security/password';
import { stableStringify } from '../../lib/security/stable-stringify';

interface AddItemPayload {
  productId?: string;
  variantSku?: string;
  qty?: number;
  source?: string;
}

interface UpdateItemPayload {
  qty?: number;
  variantSku?: string;
}

interface CheckoutPayload {
  cartId?: string;
  paymentTokenId?: string;
  shippingAddressId?: string;
  billingAddressId?: string;
  rid?: string;
  idempotencyKey?: string;
}

@Injectable()
export class CartService {
  private readonly store = inMemoryStore;

  constructor(private readonly ordersService: OrdersService) {}

  async addItem(rawPayload: Record<string, unknown>) {
    const payload = rawPayload as AddItemPayload;
    if (!payload.productId || !payload.variantSku) {
      throw new HttpException(
        { error: { code: 'CART_INVALID_PAYLOAD', message: 'productId and variantSku are required.' } },
        HttpStatus.UNPROCESSABLE_ENTITY
      );
    }

    const quantity = payload.qty ?? 1;
    if (quantity <= 0 || quantity > 5) {
      throw new HttpException(
        { error: { code: 'CART_INVALID_QUANTITY', message: 'Quantity must be between 1 and 5.' } },
        HttpStatus.UNPROCESSABLE_ENTITY
      );
    }

    const product = this.ensureProduct(payload.productId);
    const variant = this.resolveVariant(product, payload.variantSku);
    if (variant.stock < quantity) {
      throw new HttpException(
        { error: { code: 'SKU_UNAVAILABLE', message: 'Insufficient inventory for variant.' } },
        HttpStatus.CONFLICT
      );
    }

    const cart = this.ensureCart();
    const existingItem = cart.items.find((item) => item.variantId === variant.id);
    if (existingItem) {
      existingItem.quantity = Math.min(existingItem.quantity + quantity, 5);
      cart.updatedAt = timestamp();
      this.store.carts.set(cart.id, cart);
      return this.formatAddItemResponse(cart, existingItem, product, variant);
    }

    const item: CartItemRecord = {
      id: generateId(),
      productId: product.id,
      variantId: variant.id,
      variantSku: variant.sku,
      quantity
    };

    cart.items.push(item);
    cart.updatedAt = timestamp();
    this.store.carts.set(cart.id, cart);

    return this.formatAddItemResponse(cart, item, product, variant);
  }

  async updateItem(itemId: string, rawPayload: Record<string, unknown>) {
    const payload = rawPayload as UpdateItemPayload;
    const cart = this.ensureCart();
    const item = cart.items.find((record) => record.id === itemId);
    if (!item) {
      throw new HttpException(
        { error: { code: 'CART_ITEM_NOT_FOUND', message: 'Cart item not found.' } },
        HttpStatus.NOT_FOUND
      );
    }

    if (payload.qty !== undefined) {
      if (payload.qty <= 0) {
        cart.items = cart.items.filter((record) => record.id !== itemId);
      } else {
        item.quantity = Math.min(payload.qty, 5);
      }
    }

    if (payload.variantSku) {
      const product = this.store.products.get(item.productId);
      if (!product) {
        throw new HttpException(
          { error: { code: 'CART_PRODUCT_NOT_FOUND', message: 'Product not found.' } },
          HttpStatus.NOT_FOUND
        );
      }
      const variant = this.resolveVariant(product, payload.variantSku);
      item.variantId = variant.id;
      item.variantSku = variant.sku;
    }

    cart.updatedAt = timestamp();
    this.store.carts.set(cart.id, cart);
    return this.getCart();
  }

  async removeItem(itemId: string) {
    const cart = this.ensureCart();
    const initialCount = cart.items.length;
    cart.items = cart.items.filter((record) => record.id !== itemId);
    if (cart.items.length === initialCount) {
      throw new HttpException(
        { error: { code: 'CART_ITEM_NOT_FOUND', message: 'Cart item not found.' } },
        HttpStatus.NOT_FOUND
      );
    }
    cart.updatedAt = timestamp();
    this.store.carts.set(cart.id, cart);
  }

  async getCart() {
    const cart = this.ensureCart();
    return this.formatCart(cart);
  }

  async checkout(rawPayload: Record<string, unknown>) {
    const payload = rawPayload as CheckoutPayload;
    const cart = this.ensureCart(payload.cartId);

    if (!payload.paymentTokenId || !payload.shippingAddressId || !payload.billingAddressId) {
      throw new HttpException(
        { error: { code: 'CHECKOUT_INVALID', message: 'Payment token and addresses are required.' } },
        HttpStatus.UNPROCESSABLE_ENTITY
      );
    }

    const idempotencyKey = payload.idempotencyKey ?? generateId();
    const payloadHash = stableStringify({ ...payload, cartId: cart.id });
    const existingIntent = this.store.checkoutIntents.get(idempotencyKey);
    if (existingIntent) {
      if (existingIntent.payloadHash !== payloadHash) {
        throw new HttpException(
          { error: { code: 'CHECKOUT_IDEMPOTENCY_CONFLICT', message: 'Payload differs for existing idempotency key.' } },
          HttpStatus.CONFLICT
        );
      }
      const existingOrder = this.store.orders.get(existingIntent.orderId);
      if (existingOrder) {
        return {
          orderId: existingOrder.id,
          status: existingOrder.status,
          paymentIntentRef: existingIntent.paymentIntentRef,
          next: {
            tracking: null,
            brandFulfillmentEta: new Date(Date.now() + 1000 * 60 * 60 * 24 * 5).toISOString().split('T')[0]
          }
        };
      }
    }

    const checkoutResult = this.ordersService.createOrderFromCart({
      cart,
      paymentTokenId: payload.paymentTokenId,
      shippingAddressId: payload.shippingAddressId,
      billingAddressId: payload.billingAddressId,
      rid: payload.rid
    });

    const intent: CheckoutIntentRecord = {
      idempotencyKey,
      payloadHash,
      orderId: checkoutResult.orderId,
      createdAt: timestamp(),
      paymentIntentRef: checkoutResult.paymentIntentRef
    };
    this.store.checkoutIntents.set(idempotencyKey, intent);

    return checkoutResult;
  }

  private ensureCart(cartId?: string): CartRecord {
    const userId = this.ensureUser();
    if (cartId) {
      const cart = this.store.carts.get(cartId);
      if (!cart) {
        throw new HttpException(
          { error: { code: 'CART_NOT_FOUND', message: 'Cart not found.' } },
          HttpStatus.NOT_FOUND
        );
      }
      return cart;
    }

    let cart = Array.from(this.store.carts.values()).find((record) => record.userId === userId);
    if (!cart) {
      cart = {
        id: generateId(),
        userId,
        items: [],
        updatedAt: timestamp()
      };
      this.store.carts.set(cart.id, cart);
    }
    return cart;
  }

  private ensureProduct(productId: string): ProductRecord {
    const product = this.store.products.get(productId);
    if (product) {
      return product;
    }

    const brandId = generateId();
    this.store.brands.set(brandId, {
      id: brandId,
      name: 'Core Fit Brand',
      slug: 'core-fit',
      onboarded: true,
      createdAt: timestamp(),
      updatedAt: timestamp()
    });

    const record: ProductRecord = {
      id: productId,
      brandId,
      name: 'Core Fit Tee',
      description: 'Lightweight tee optimized for FitTwin try-on.',
      heroImageUrl: 'https://cdn.fittwin.local/products/core-fit-tee.png',
      active: true,
      sizeChartId: undefined,
      fitMapId: undefined,
      createdAt: timestamp(),
      updatedAt: timestamp()
    };
    this.store.products.set(productId, record);

    const variant: ProductVariantRecord = {
      id: generateId(),
      productId,
      sku: 'FT-TEE-M',
      label: 'M',
      attributes: { chest: 100, waist: 82 },
      stock: 50,
      priceCents: 4200,
      currency: 'USD',
      createdAt: timestamp(),
      updatedAt: timestamp()
    };
    this.store.variants.set(variant.id, variant);

    const altVariant: ProductVariantRecord = {
      id: generateId(),
      productId,
      sku: 'FT-TEE-L',
      label: 'L',
      attributes: { chest: 106, waist: 88 },
      stock: 30,
      priceCents: 4200,
      currency: 'USD',
      createdAt: timestamp(),
      updatedAt: timestamp()
    };
    this.store.variants.set(altVariant.id, altVariant);

    return record;
  }

  private resolveVariant(product: ProductRecord, variantSku: string): ProductVariantRecord {
    const variants = Array.from(this.store.variants.values()).filter((variant) => variant.productId === product.id);
    const match = variants.find((variant) => variant.sku === variantSku);
    if (!match) {
      throw new HttpException(
        { error: { code: 'SKU_UNAVAILABLE', message: 'Variant SKU not recognized for product.' } },
        HttpStatus.UNPROCESSABLE_ENTITY
      );
    }
    return match;
  }

  private ensureUser(): string {
    const existingUser = Array.from(this.store.users.values())[0];
    if (existingUser) {
      return existingUser.id;
    }

    const email = 'demo@fittwin.local';
    const userId = generateId();
    const now = timestamp();
    const { hash, salt } = hashPassword('DemoPass!123');

    this.store.users.set(userId, {
      id: userId,
      email,
      passwordHash: hash,
      passwordSalt: salt,
      status: 'active',
      consent: { terms: true, marketing: false, privacy: true },
      createdAt: now,
      updatedAt: now,
      failedAttempts: 0
    });

    this.store.profiles.set(userId, {
      userId,
      email,
      username: 'fitfan',
      appearance: {},
      stylePreferences: {},
      bodyMetrics: {},
      consents: { marketing: false, dataExportAvailable: false },
      avatars: [],
      updatedAt: now
    });

    return userId;
  }

  private formatCart(cart: CartRecord) {
    const items = cart.items.map((item) => {
      const product = this.store.products.get(item.productId);
      const variant = this.store.variants.get(item.variantId);
      return {
        itemId: item.id,
        productId: item.productId,
        variantSku: item.variantSku,
        name: product?.name ?? 'Unknown product',
        sizeLabel: variant?.label ?? 'N/A',
        qty: item.quantity,
        unitPrice: variant?.priceCents ?? 0,
        currency: variant?.currency ?? 'USD',
        recommended: true,
        fitSummary: { confidence: 88, notes: ['waist snug'] }
      };
    });

    const subtotal = items.reduce((sum, item) => sum + (item.unitPrice ?? 0) * item.qty, 0);
    const tax = Math.round(subtotal * 0.0825);
    const shipping = subtotal > 10000 ? 0 : 1200;

    return {
      cartId: cart.id,
      items,
      totals: {
        subtotal,
        shipping,
        tax,
        currency: 'USD'
      },
      recommendations: items.slice(0, 1).map((item) => ({ productId: item.productId, sizeRec: item.sizeLabel }))
    };
  }

  private formatAddItemResponse(
    cart: CartRecord,
    item: CartItemRecord,
    product: ProductRecord,
    variant: ProductVariantRecord
  ) {
    return {
      cartId: cart.id,
      itemId: item.id,
      item: {
        productId: product.id,
        variantSku: variant.sku,
        name: product.name,
        sizeLabel: variant.label,
        qty: item.quantity,
        unitPrice: variant.priceCents,
        currency: variant.currency,
        recommended: true,
        fitSummary: { confidence: 88, notes: ['waist snug'] }
      }
    };
  }
}

File: services/python/measurement/backend/app/services/cart_service.py
"""
Cart Service

Handles cart management including adding/updating/removing items,
inventory validation, and cart persistence.
"""

from typing import Dict, List, Optional, Any
from datetime import datetime
from supabase import Client
import os


class CartService:
    """Service for managing shopping carts."""

    MAX_QUANTITY_PER_ITEM = 5
    FREE_SHIPPING_THRESHOLD = 10000  # cents

    def __init__(self, supabase_client: Client):
        """Initialize cart service with Supabase client."""
        self.db = supabase_client

    async def get_cart(self, user_id: str) -> Dict[str, Any]:
        """
        Get the current cart for a user.

        Args:
            user_id: User ID

        Returns:
            Cart with items and totals
        """
        # Get or create cart
        cart = await self._ensure_cart(user_id)

        # Get cart items with product and variant details
        items_response = self.db.table("cart_items")\
            .select("*, products(*), product_variants(*)")\
            .eq("cart_id", cart["id"])\
            .execute()

        items = []
        subtotal = 0

        for item in items_response.data:
            variant = item["product_variants"]
            product = item["products"]

            unit_price = variant["price_cents"]
            line_total = unit_price * item["quantity"]
            subtotal += line_total

            items.append({
                "item_id": item["id"],
                "product_id": product["id"],
                "variant_sku": variant["sku"],
                "name": product["name"],
                "size_label": variant["label"],
                "qty": item["quantity"],
                "unit_price": unit_price,
                "currency": variant["currency"],
                "recommended": item.get("recommended", False),
                "fit_summary": item.get("fit_summary", {})
            })

        # Calculate totals
        tax = int(subtotal * 0.0825)  # 8.25% tax rate
        shipping = 0 if subtotal >= self.FREE_SHIPPING_THRESHOLD else 1200

        return {
            "cart_id": cart["id"],
            "items": items,
            "totals": {
                "subtotal": subtotal,
                "shipping": shipping,
                "tax": tax,
                "total": subtotal + shipping + tax,
                "currency": "USD"
            }
        }

    async def add_item(
        self,
        user_id: str,
        product_id: str,
        variant_sku: str,
        quantity: int = 1,
        recommended: bool = False,
        fit_summary: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """
        Add an item to the cart.

        Args:
            user_id: User ID
            product_id: Product ID
            variant_sku: Variant SKU
            quantity: Quantity to add (1-5)
            recommended: Whether this was a recommended size
            fit_summary: Fit analysis data

        Returns:
            Updated cart item

        Raises:
            ValueError: If quantity is invalid or inventory insufficient
        """
        # Validate quantity
        if quantity <= 0 or quantity > self.MAX_QUANTITY_PER_ITEM:
            raise ValueError(f"Quantity must be between 1 and {self.MAX_QUANTITY_PER_ITEM}")

        # Get variant and check inventory
        variant_response = self.db.table("product_variants")\
            .select("*")\
            .eq("sku", variant_sku)\
            .eq("product_id", product_id)\
            .single()\
            .execute()

        if not variant_response.data:
            raise ValueError("Variant not found")

        variant = variant_response.data

        if variant["stock"] < quantity:
            raise ValueError("Insufficient inventory")

        # Get or create cart
        cart = await self._ensure_cart(user_id)

        # Check if item already exists in cart
        existing_item_response = self.db.table("cart_items")\
            .select("*")\
            .eq("cart_id", cart["id"])\
            .eq("variant_id", variant["id"])\
            .execute()

        if existing_item_response.data:
            # Update existing item
            existing_item = existing_item_response.data[0]
            new_quantity = min(
                existing_item["quantity"] + quantity,
                self.MAX_QUANTITY_PER_ITEM
            )

            updated_item = self.db.table("cart_items")\
                .update({
                    "quantity": new_quantity,
                    "updated_at": datetime.utcnow().isoformat()
                })\
                .eq("id", existing_item["id"])\
                .execute()

            return await self._format_cart_item(updated_item.data[0])
        else:
            # Create new cart item
            new_item = self.db.table("cart_items")\
                .insert({
                    "cart_id": cart["id"],
                    "product_id": product_id,
                    "variant_id": variant["id"],
                    "quantity": quantity,
                    "recommended": recommended,
                    "fit_summary": fit_summary or {}
                })\
                .execute()

            return await self._format_cart_item(new_item.data[0])

    async def update_item(
        self,
        user_id: str,
        item_id: str,
        quantity: Optional[int] = None,
        variant_sku: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Update a cart item.

        Args:
            user_id: User ID
            item_id: Cart item ID
            quantity: New quantity (if None, keep current)
            variant_sku: New variant SKU (if None, keep current)

        Returns:
            Updated cart

        Raises:
            ValueError: If item not found or invalid update
        """
        cart = await self._ensure_cart(user_id)

        # Get cart item
        item_response = self.db.table("cart_items")\
            .select("*")\
            .eq("id", item_id)\
            .eq("cart_id", cart["id"])\
            .single()\
            .execute()

        if not item_response.data:
            raise ValueError("Cart item not found")

        update_data = {"updated_at": datetime.utcnow().isoformat()}

        # Update quantity
        if quantity is not None:
            if quantity <= 0:
                # Remove item if quantity is 0 or negative
                return await self.remove_item(user_id, item_id)
            else:
                update_data["quantity"] = min(quantity, self.MAX_QUANTITY_PER_ITEM)

        # Update variant
        if variant_sku:
            variant_response = self.db.table("product_variants")\
                .select("*")\
                .eq("sku", variant_sku)\
                .single()\
                .execute()

            if not variant_response.data:
                raise ValueError("Variant not found")

            update_data["variant_id"] = variant_response.data["id"]

        # Apply update
        self.db.table("cart_items")\
            .update(update_data)\
            .eq("id", item_id)\
            .execute()

        return await self.get_cart(user_id)

    async def remove_item(self, user_id: str, item_id: str) -> None:
        """
        Remove an item from the cart.

        Args:
            user_id: User ID
            item_id: Cart item ID

        Raises:
            ValueError: If item not found
        """
        cart = await self._ensure_cart(user_id)

        result = self.db.table("cart_items")\
            .delete()\
            .eq("id", item_id)\
            .eq("cart_id", cart["id"])\
            .execute()

        if not result.data:
            raise ValueError("Cart item not found")

    async def clear_cart(self, user_id: str) -> None:
        """
        Clear all items from the cart.

        Args:
            user_id: User ID
        """
        cart = await self._ensure_cart(user_id)

        self.db.table("cart_items")\
            .delete()\
            .eq("cart_id", cart["id"])\
            .execute()

    async def _ensure_cart(self, user_id: str) -> Dict[str, Any]:
        """
        Get or create a cart for the user.

        Args:
            user_id: User ID

        Returns:
            Cart record
        """
        # Try to get existing cart
        cart_response = self.db.table("carts")\
            .select("*")\
            .eq("user_id", user_id)\
            .execute()

        if cart_response.data:
            return cart_response.data[0]

        # Create new cart
        new_cart = self.db.table("carts")\
            .insert({"user_id": user_id})\
            .execute()

        return new_cart.data[0]

    async def _format_cart_item(self, item: Dict[str, Any]) -> Dict[str, Any]:
        """Format a cart item with product and variant details."""
        # Get product and variant details
        variant = self.db.table("product_variants")\
            .select("*")\
            .eq("id", item["variant_id"])\
            .single()\
            .execute().data

        product = self.db.table("products")\
            .select("*")\
            .eq("id", item["product_id"])\
            .single()\
            .execute().data

        return {
            "item_id": item["id"],
            "product_id": product["id"],
            "variant_sku": variant["sku"],
            "name": product["name"],
            "size_label": variant["label"],
            "qty": item["quantity"],
            "unit_price": variant["price_cents"],
            "currency": variant["currency"],
            "recommended": item.get("recommended", False),
            "fit_summary": item.get("fit_summary", {})
        }
